sound/soc/sof/ipc4.c

```diff
@@ -352,6 +352,9 @@ static int ipc4_tx_msg_unlocked(struct snd_sof_ipc *ipc,
 		return ret;
 	}
 
+	if (ipc4_msg->callback_after_send)
+		ipc4_msg->callback_after_send(sdev, ipc4_msg->callback_data);
```

Comment:
"We cannot invoke the callback AFTER the IPC response as the response will not come unless we execute the callback after sending the message."

That's the part you have not described at all in any of your commit messages or comments.
This is a special case of an IPC needing an extra configuration before it can reply.

"callback_after_send" does not even begin to describe what you are trying to do. "after_send" could be understood as "after pushing the IPC button, before the reply" or "after completing an IPC, reply included".

Reply:
OK, I will try to reword the commit messages and the comment for the `callback_after_send`, for which I don't have a better name in my mind atm.

Reply:
"callback_before_ipc_reply" ? It's literally what the design intent is.